### PR TITLE
Make document serializable, create utility to create a docstore

### PR DESCRIPTION
**Pull request by eyurtsev**
This PR makes the following changes:

1. Documents become serializable using langhchain serialization
2. Make a utility to create a docstore kw store

Will help to address issue here: https://github.com/langchain-ai/langchain/issues/9345

### CODE REVIEW

**Comment by eyurtsev** [comment]
@baskaryan / @hwchase17 let me know if you agree with namespacing, and if code looks correct.

Largest concern is that a key-value wrapper that allows writing and fetching documents != a doc-store in terms of obvious features

File: libs/langchain/langchain/schema/document.py [hunk 1]
```diff
@@ -17,6 +17,16 @@ class Document(Serializable):
         documents, etc.).
     """
 
+    @property
+    def lc_serializable(self) -> bool:
+        """Return whether or not the class is serializable."""
+        return True
+
+    @property
+    def lc_namespace(self) -> List[str]:
```
Comment (nfcampos) [inline libs/langchain/langchain/schema/document.py:26]: This is not necessary, the only reason you find this in runnables is to maintain backwards compat with the folder they previously lived in

File: libs/langchain/langchain/docstore/_kw_storage.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    Args:
+        store: The base store to wrap.
+
+    Returns:
+        A key-value store for documents.
+    """
+    return EncoderBackedStore(
+        store=store,
+        key_encoder=lambda key: key,
```
Comment (nfcampos) [inline libs/langchain/langchain/docstore/_kw_storage.py:35]: Should this be configurable as an arg?

**Review by nfcampos — APPROVED** [review]

File: libs/langchain/langchain/storage/_lc_store.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    return EncoderBackedStore(
+        store,
+        key_encoder or _identity,
+        _dump_as_bytes,
+        _load_from_bytes,
+    )
+
+
+def create_kv_docstore(
```
Comment (eyurtsev) [inline libs/langchain/langchain/storage/_lc_store.py:66]: Other possibility below -- but seems like over abstraction

Objects we'll want to serialize are:

1. everything (then no need to type check)
2. prompts (can do special purpose)
3. documents (can do special purpose)

----

The other possibility if anyone has time is to try and introduce a more general class that proxies the embedding class but gets the type signatures right


The tricky thing is getting the type signatures right :) 

user passes `types` and the class does run time type checking.



```


class LCStore(BaseStore[str, T]):
    def __init__(self, store: BaseStore[str, bytes], types: Sequence[Type]):
        """Create a store for langchain serializable objects from a bytes store."""
        self.types = types
        self.store = EncoderBackedStore(
            store,
            key_encoder or _identity,
            _dump_document_as_bytes,
            _load_document_from_bytes,
        )

    def mget(self, keys: Sequence[K]) -> List[Optional[V]]:
        """Get multiple keys."""
        return self.store.mget(keys)

    def mdelete(self, keys: Sequence[K]) -> None:
        """Delete multiple keys."""
        return self.store.mdelete(keys)

    def yield_keys(
        self, *, prefix: Optional[str] = None
    ) -> Union[Iterator[K], Iterator[str]]:
        """Yield all keys in the store."""
        yield from self.store.yield_keys(prefix=prefix)

    def mset(self, items: Sequence[Tuple[K, V]]) -> None:
        """Set multiple key-value pairs."""
        return self.store.mset(items)
```

File: libs/langchain/langchain/storage/_lc_store.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    """Return a bytes representation of a document."""
+    if not isinstance(obj, Document):
+        raise TypeError("Expected a Document instance")
+    return dumps(obj).encode("utf-8")
+
+
+def _load_document_from_bytes(serialized: bytes) -> Document:
+    """Return a document from a bytes representation."""
+    obj = loads(serialized.decode("utf-8"))
+    if not isinstance(obj, Document):
```
Comment (nfcampos) [inline libs/langchain/langchain/storage/_lc_store.py:26]: So all that is different about the specialised dump/load functions is the isinstance check?

Comment (eyurtsev) [reply libs/langchain/langchain/storage/_lc_store.py:26]: Yeah run time check and type annotation

**Pull request merged**